instance/models/utils.py

```diff
@@ -44,3 +55,263 @@ def save(self, *args, **kwargs):
+    def __eq__(self, obj):
+        """
+        Syntactic sugar to make comparing states easier.
+
+        obj should be a ResourceState class or instance.
+        """
+        try:
+            return issubclass(self.__class__, obj)
+        except TypeError:  # The provided argument is not a class
+            return type(obj) is type(self)
+
+    @classmethod
+    def one_of(cls, *state_classes):
+        """ Syntactic sugar to make comparing an instance to a set of types easier """
+        return issubclass(cls, state_classes)
+
+    def __init__(self, resource, state_manager):
+        """ Instantiate this state, saving a reference to the parent resource """
+        self._resource = resource
+        self._state_manager = state_manager
+
+    class Enum:
+        """
+        Syntacic sugar for declaring a group of ResourceState classes inside a class.
+
+        Use like:
+
+        class MyStates(ResourceState.Enum):
+            class State1(ResourceState):
+                state_id = '1'
+            class State2(ResourceState):
+                state_id = '2'
+
+        or if states are already declared, use like:
+        class MyStates(ResourceState.Enum):
+            State1, State2 = State1, State2
+
+        Then you can use MyStates.states to get a list of the state classes.
+        """
+        @ClassProperty
+        @classmethod
+        def states(cls):
+            """
+            Get an iterable listing all the classes defined within this class
+            """
+            return tuple(state for name, state in cls.__dict__.items() if name[:1] != '_' and name != 'states')
+
+
+class ResourceStateDescriptor:
+    """
+    Descriptor which implements a finite state machine.
+
+    Assign an instance of this to a class attribute such as 'state'. This will create a 'state'
+    property that always returns an instance of one of the allowed state_classes.
+    The 'state' property cannot be assigned to; only the current state instance is allowed to
+    change the state. This makes it easy to reason about the behavior of the state.
+    """
+    def __init__(self, state_classes, default_state, model_field_name=None):
+        """
+        Instantiate a ResourceStateDescriptor to manage a state machine.
+
+        state_classes: A list of class types that are valid states for this state machine.
+        default_state: If no state has been set, assume the state is this state class.
+        model_field_name: The name of a django CharField to keep updated with the name of the
+            current state, or None. (caching the state in a field is optional)
+        """
+        self.state_classes = state_classes
+        assert len(set(state_class.state_id for state_class in state_classes)) == len(state_classes), \
+            "A resource's states must each have a unique state_id"
+        assert default_state in state_classes
+        self.default_state_class = default_state
+        self.model_field_name = model_field_name
```

Comment:
I like that idea, though currently every usage we have uses model_field_name. I didn't get time to do this refactoring today, but I do like it in principle.